variables.tf

```diff
@@ -1,9 +1,444 @@
+  validation {
+    condition     = contains(["resource", "serviceInstance", "resourceGroup"], var.policy_granularity)
+    error_message = "Invalid policy_granularity: use one of \"resource\", \"serviceInstance\", or \"resourceGroup\""
+  }
+}
+
+##############################################################################
+# Common variables
+##############################################################################
+
+variable "create_resource_group" {
+  type        = bool
+  description = "Whether the value of var.resource_group_name should be a new of existing resource_group"
+  default     = false
+}
+
+variable "resource_group_name" {
+  type        = string
+  description = "The name of an existing resource group to provision resources in to. If not set a new resource group will be created using the prefix variable"
+  default     = "Default"
+
+  validation {
+    condition     = can(regex("^[a-zA-Z0-9-_ ]+$", var.resource_group_name)) && length(var.resource_group_name) < 40
+    error_message = "Must be a valid resource group name that matches ^[a-zA-Z0-9-_ ]+$ and be less than 40 characters"
+  }
+}
+
+variable "resource_tags" {
+  type        = list(string)
+  description = "Optional list of tags to be added to created resources"
+  default     = []
+}
+
+variable "access_tags" {
+  type        = list(string)
+  description = "A list of access tags to apply to the cos instance created by the module, see https://cloud.ibm.com/docs/account?topic=account-access-tags-tutorial for more details"
+  default     = []
+
+  validation {
+    condition = alltrue([
+      for tag in var.access_tags : can(regex("[\\w\\-_\\.]+:[\\w\\-_\\.]+", tag)) && length(tag) <= 128
+    ])
+    error_message = "Tags must match the regular expression \"[\\w\\-_\\.]+:[\\w\\-_\\.]+\", see https://cloud.ibm.com/docs/account?topic=account-tag&interface=ui#limits for more details"
+  }
+}
+
+# region needs to provide cross region support.
+variable "region" {
+  description = "Region where resources will be created"
+  type        = string
+  default     = "us-south"
+
+  validation {
+    condition     = can(regex("us-south|eu-de|jp-tok", var.region))
+    error_message = "Variable 'region' must be 'us-south' or 'eu-de', or 'jp-tok'. The encryption key must be created in a high availability key protect instance for cross region object storage. See https://cloud.ibm.com/docs/key-protect?topic=key-protect-ha-dr"
+  }
+}
+
 ########################################################################################################################
-# Input Variables
+# Key_protect
 ########################################################################################################################
 
-#variable "my_variable" {
-#  type        = string
-#  description = "A description of my variable"
-#  default     = "default_value"
-#}
+variable "create_key_protect_instance" {
+  type        = bool
+  description = "Key Protect instance name"
```

Comment:
I suggest to mention the existing KMS instance input variable here, and that it would be a required input if this value is set to false